libs/common/AsciiUtils.cs

```diff
@@ -1,17 +1,87 @@
 ﻿// Copyright (c) Microsoft Corporation.
 // Licensed under the MIT license.
 
+using System;
+using System.Diagnostics;
+using System.Text;
+
 namespace Garnet.common;
 
 /// <summary>
 /// Utilites for ASCII parsing and manipulation.
 /// </summary>
+/// <remarks>
+/// This class polyfills various <see cref="char"/> and <c>Ascii</c> methods for .NET 6.
+/// </remarks>
 public static class AsciiUtils
 {
-    public static byte ToLower(byte value)
+    /// <summary>Indicates whether a character is within the specified inclusive range.</summary>
+    /// <param name="c">The character to evaluate.</param>
+    /// <param name="minInclusive">The lower bound, inclusive.</param>
+    /// <param name="maxInclusive">The upper bound, inclusive.</param>
+    /// <returns>true if <paramref name="c"/> is within the specified range; otherwise, false.</returns>
+    /// <remarks>
+    /// The method does not validate that <paramref name="maxInclusive"/> is greater than or equal
+    /// to <paramref name="minInclusive"/>.  If <paramref name="maxInclusive"/> is less than
+    /// <paramref name="minInclusive"/>, the behavior is undefined.
+    /// </remarks>
+    public static bool IsBetween(byte c, char minInclusive, char maxInclusive)
+    {
+        return (uint)(c - minInclusive) <= (uint)(maxInclusive - minInclusive);
+    }
+
+    public static byte ToLower(byte c)
+    {
+        if (IsBetween(c, 'A', 'Z'))
+            c = (byte)(c | 0x20);
+        return c;
+    }
+    public static byte ToUpper(byte c)
+    {
+        if (IsBetween(c, 'a', 'z'))
+            c = (byte)(c & ~0x20);
+        return c;
+    }
+
+    /// <summary>
+    /// Convert ASCII Span to upper case
+    /// </summary>
+    public static void ToUpperInPlace(Span<byte> command)
     {
-        if ((uint)(value - 'A') <= (uint)('Z' - 'A')) // Is in [A-Z]
-            value = (byte)(value | 0x20);
-        return value;
+#if NET8_0_OR_GREATER
+        Ascii.ToUpperInPlace(command, out _);
+#else
+        foreach (ref var c in command)
+            if (c > 96 && c < 123)
+                c -= 32;
+#endif
+    }
+
+    /// <summary>
+    /// Check if two byte spans are equal, where right is an all-upper-case span, ignoring case if there are ASCII bytes.
+    /// </summary>
+    public static bool EqualsIgnoreCase(ReadOnlySpan<byte> left, ReadOnlySpan<byte> right)
+    {
+#if NET8_0_OR_GREATER
+        return Ascii.EqualsIgnoreCase(left, right);
```

Comment:
The idea of `EqualsUpperCaseSpanIgnoringCase` is that we assume the common case is both are upper case in which Sequence.Equals is the fast path. Only otherwise do we pay the detailed comparison cost. Also, we do not want to call Ascii.EqualsIgnoreCase even in the slow path, as our loop knows that the right side is upper-case and thus can avoid a few comparisons.

Reply:
Even the vector optimized version [here](https://github.com/gfoidl/dotnet-runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Text/Ascii.Equality.cs#L195) has to do a lot as it is unable to make assumptions about ASCII or the right being upper case.

Reply:
Alright then, I don't feel strongly about this one. 

We will get JIT to completely unroll the half-const inputs someday though :) I have been pestering one of the JIT developers about it for a while now and thought to use this as motivation 😆